modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/QueryExecutionProgram.java

```diff
@@ -75,6 +80,32 @@ private QueryExecutionProgram() {
     static boolean errorHandler(Query query, Throwable th) {
         if (canRecover(query, th)) {
             query.error.set(null);
+
+            if (multiStepPlanOutdated(th)) {
+                assert query.plan instanceof MultiStepPlan;
+                assert query.currentPhase() == ExecutionPhase.CURSOR_INITIALIZATION : query.currentPhase();
+
+                SqlOperationContext context = query.operationContext;
+                QueryTransactionWrapper txWrapper = query.usedTransaction;
+
+                assert context != null;
+                assert txWrapper != null;
+
+                context.getAndSetTxOnRetry(txWrapper);
```

Comment:
currently, error handling for this case is scattered across number of places. I would propose to move all preparation here. Looks like we just need to recreate operationContext with all necessary changes (operation time, transaction to use, if any).

Reply:
Thanks, fixed, but with some caveats:
1. The context is recreated (copied) via a static method of `SqlOperationContext.withTransactionForRetry(tx)`, since there's currently no external access to, for example, `errorListener` .
2. The method to retrieve transaction on retry (`SqlOperationContext.retryTx()`) has getAndSet(null) semantic to avoid issues with retry/recovery AFTER re-planning the query.